File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Sprite.Painting/GameComponentCollection.cs
using System.Collections.Generic;
using Loon.Utils;
using Loon.Java.Collections;
using Loon.Core.Timer;
namespace Loon.Action.Sprite.Painting {
	
	public sealed class GameComponentCollection {

        public sealed class GameComponent_0 : IComparer<IGameComponent>
        {
			public int Compare(IGameComponent one, IGameComponent two) {
				if (one  is  DrawableGameComponent
						&& two  is  DrawableGameComponent) {
					return ((DrawableGameComponent) one).GetDrawOrder()
							- ((DrawableGameComponent) two).GetDrawOrder();
				}
				return 0;
			}
		}

        public sealed class GameComponent_1 : IComparer<IGameComponent>
        {
			public int Compare(IGameComponent one, IGameComponent two) {
				if (one  is  GameComponent && two  is  GameComponent) {
					return ((GameComponent) one).GetUpdateOrder()
							- ((GameComponent) two).GetUpdateOrder();
				}
				return 0;
			}
		}
	
		private List<IGameComponent> collections;

        private List<IGameComponent> collectionsToUpdate;

        private List<IGameComponent> collectionsToDraw;

        private IComparer<IGameComponent> igameDrawComparator = new GameComponentCollection.GameComponent_0();

        private IComparer<IGameComponent> igameUpdateComparator = new GameComponentCollection.GameComponent_1();
	
		public GameComponentCollection() {
			this.collections = new List<IGameComponent>();
			this.collectionsToUpdate = new List<IGameComponent>();
			this.collectionsToDraw = new List<IGameComponent>();
		}
	
		public int Size() {
			return collections.Count;
		}
	
		public void SortDraw() {
			 Arrays.Sort(collections,igameDrawComparator);
		}
	
		public void SortUpdate() {
            Arrays.Sort(collections, igameUpdateComparator);
		}
	
		public List<IGameComponent> List() {
			return new List<IGameComponent>(collections);
		}
	
		public IGameComponent Get(int idx) {
			return collections[idx];
		}
	
		internal void Draw(SpriteBatch batch, GameTime gameTime) {
			if (isClear) {
				return;
			}
			if (collectionsToDraw.Count > 0) {
				CollectionUtils.Clear(collectionsToDraw);
			}
			foreach (IGameComponent drawable  in  collections) {
				CollectionUtils.Add(collectionsToDraw,drawable);
			}
			foreach (IGameComponent drawable  in  collectionsToDraw) {
				if (drawable  is  IDrawable) {
					IDrawable comp = (IDrawable) drawable;
					comp.Draw(batch, gameTime);
				}
			}
		}
	
		internal void Load() {
			isClear = false;
			foreach (IGameComponent comp  in  collections) {
				comp.Initialize();
			}
		}
	
		internal void Update(GameTime gameTime) {
			if (isClear) {
				return;
			}
			if (collectionsToUpdate.Count > 0) {
				CollectionUtils.Clear(collectionsToUpdate);
			}
			foreach (IGameComponent drawable  in  collections) {
				CollectionUtils.Add(collectionsToUpdate,drawable);
			}
            IGameComponent _drawable;
			int screenIndex;
			for (; collectionsToUpdate.Count > 0;) {
	
				screenIndex = collectionsToUpdate.Count - 1;
                _drawable = collectionsToUpdate[screenIndex];
	
				CollectionUtils.RemoveAt(collectionsToUpdate,screenIndex);

                if (_drawable is IUpdateable)
                {
                    IUpdateable comp = (IUpdateable)_drawable;
					comp.Update(gameTime);
				}
			}
	
		}
	
		private bool isClear;
	
		public void Clear() {
			if (!isClear) {
				 lock (typeof(GameComponentCollection)) {
								CollectionUtils.Clear(collections);
								CollectionUtils.Clear(collectionsToUpdate);
								CollectionUtils.Clear(collectionsToDraw);
								isClear = true;
							}
			}
		}
	
		public bool Add(IGameComponent gc) {
			if (isClear) {
				return false;
			}
			gc.Initialize();
			bool result = CollectionUtils.Add(collections,gc);
			if (gc != null && Added != null) {
				Added.Invoke(gc);
			}
			if (gc  is  DrawableGameComponent) {
				if (((DrawableGameComponent) gc).GetDrawOrder() != 0) {
					SortDraw();
				}
			}
			return result;
		}
	
		public bool Add(IGameComponent gc, int index) {
			if (isClear) {
				return false;
			}
			gc.Initialize();
			bool result = CollectionUtils.Add(collections,gc);
			for (int i = 0; i < collections.Count; i++) {
				if (collections[i]   is  DrawableGameComponent) {
					if (i == index) {
						((DrawableGameComponent) collections[i])
								.SetEnabled(true);
					} else {
						((DrawableGameComponent) collections[i])
								.SetEnabled(false);
					}
				}
			}
			if (gc != null && Added != null) {
				Added.Invoke(gc);
			}
			if (gc  is  DrawableGameComponent) {
				if (((DrawableGameComponent) gc).GetDrawOrder() != 0) {
					SortDraw();
				}
			}
			return result;
		}
	
		public bool Remove(IGameComponent gc) {
			if (isClear) {
				return false;
			}
			bool result = false;
			if (gc != null && gc  is  DrawableGameComponent) {
				DrawableGameComponent comp = (DrawableGameComponent) gc;
				comp.UnloadContent();
				result = CollectionUtils.Remove(collections,gc);
				CollectionUtils.Remove(collectionsToUpdate,gc);
				if (Removed != null) {
					Removed.Invoke(gc);
				}
			}
			return result;
		}
	
		public bool RemoveAt(int idx) {
			if (isClear) {
				return false;
			}
            IGameComponent comp = (IGameComponent)CollectionUtils.RemoveAt(collections, idx);
			bool result = (comp != null);
			if (result) {
				if (comp  is  DrawableGameComponent) {
					((DrawableGameComponent) comp).UnloadContent();
				}
				CollectionUtils.Remove(collectionsToUpdate,comp);
				if (Removed != null) {
					Removed.Invoke(comp);
				}
			}
			return result;
		}
	
		private ComponentEvent Added;
	
		private ComponentEvent Removed;
	
		public ComponentEvent GetAdded() {
			return Added;
		}
	
		public void SetAdded(ComponentEvent added) {
			Added = added;
		}
	
		public ComponentEvent GetRemoved() {
			return Removed;
		}
	
		public void SetRemoved(ComponentEvent removed) {
			Removed = removed;
		}
	}
}

File: Java/old/OpenGL-1.0(old_ver)/Loon-backend-Android/src/loon/action/sprite/painting/GameComponentCollection.java
/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
package loon.action.sprite.painting;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import loon.action.sprite.SpriteBatch;
import loon.core.timer.GameTime;

public final class GameComponentCollection {

	private ArrayList<IGameComponent> collections;

	private ArrayList<IGameComponent> collectionsToUpdate;

	private ArrayList<IGameComponent> collectionsToDraw;

	private Comparator<IGameComponent> igameDrawComparator = new Comparator<IGameComponent>() {
		@Override
		public int compare(IGameComponent one, IGameComponent two) {
			if (one instanceof DrawableGameComponent
					&& two instanceof DrawableGameComponent) {
				return ((DrawableGameComponent) one).getDrawOrder()
						- ((DrawableGameComponent) two).getDrawOrder();
			}
			return 0;
		}
	};

	private Comparator<IGameComponent> igameUpdateComparator = new Comparator<IGameComponent>() {
		@Override
		public int compare(IGameComponent one, IGameComponent two) {
			if (one instanceof GameComponent && two instanceof GameComponent) {
				return ((GameComponent) one).getUpdateOrder()
						- ((GameComponent) two).getUpdateOrder();
			}
			return 0;
		}
	};

	public GameComponentCollection() {
		this.collections = new ArrayList<IGameComponent>();
		this.collectionsToUpdate = new ArrayList<IGameComponent>();
		this.collectionsToDraw = new ArrayList<IGameComponent>();
	}

	public int size() {
		return collections.size();
	}

	public void sortDraw() {
		Collections.sort(collections, igameDrawComparator);
	}

	public void sortUpdate() {
		Collections.sort(collections, igameUpdateComparator);
	}

	public ArrayList<IGameComponent> list() {
		return new ArrayList<IGameComponent>(collections);
	}

	public IGameComponent get(int idx) {
		return collections.get(idx);
	}

	void draw(SpriteBatch batch, GameTime gameTime) {
		if (isClear) {
			return;
		}
		if (collectionsToDraw.size() > 0) {
			collectionsToDraw.clear();
		}
		for (IGameComponent drawable : collections) {
			collectionsToDraw.add(drawable);
		}
		for (IGameComponent drawable : collectionsToDraw) {
			if (drawable instanceof IDrawable) {
				IDrawable comp = (IDrawable) drawable;
				comp.draw(batch, gameTime);
			}
		}
	}

	void load() {
		isClear = false;
		for (IGameComponent comp : collections) {
			comp.initialize();
		}
	}

	void update(GameTime gameTime) {
		if (isClear) {
			return;
		}
		if (collectionsToUpdate.size() > 0) {
			collectionsToUpdate.clear();
		}
		for (IGameComponent drawable : collections) {
			collectionsToUpdate.add(drawable);
		}

		IGameComponent drawable;
		int screenIndex;
		for (; collectionsToUpdate.size() > 0;) {

			screenIndex = collectionsToUpdate.size() - 1;
			drawable = collectionsToUpdate.get(screenIndex);

			collectionsToUpdate.remove(screenIndex);

			if (drawable instanceof IUpdateable) {
				IUpdateable comp = (IUpdateable) drawable;
				comp.update(gameTime);
			}
		}

	}

	private boolean isClear;

	public void clear() {
		if (!isClear) {
			synchronized (GameComponentCollection.class) {
				collections.clear();
				collectionsToUpdate.clear();
				collectionsToDraw.clear();
				isClear = true;
			}
		}
	}

	public boolean add(IGameComponent gc) {
		if (isClear) {
			return false;
		}
		gc.initialize();
		boolean result = collections.add(gc);
		if (gc != null && Added != null) {
			Added.invoke(gc);
		}
		if (gc instanceof DrawableGameComponent) {
			if (((DrawableGameComponent) gc).getDrawOrder() != 0) {
				sortDraw();
			}
		}
		return result;
	}

	public boolean add(IGameComponent gc, int index) {
		if (isClear) {
			return false;
		}
		gc.initialize();
		boolean result = collections.add(gc);
		for (int i = 0; i < collections.size(); i++) {
			if (collections.get(i) instanceof DrawableGameComponent) {
				if (i == index) {
					((DrawableGameComponent) collections.get(i))
							.setEnabled(true);
				} else {
					((DrawableGameComponent) collections.get(i))
							.setEnabled(false);
				}
			}
		}
		if (gc != null && Added != null) {
			Added.invoke(gc);
		}
		if (gc instanceof DrawableGameComponent) {
			if (((DrawableGameComponent) gc).getDrawOrder() != 0) {
				sortDraw();
			}
		}
		return result;
	}

	public boolean remove(IGameComponent gc) {
		if (isClear) {
			return false;
		}
		boolean result = false;
		if (gc != null && gc instanceof DrawableGameComponent) {
			DrawableGameComponent comp = (DrawableGameComponent) gc;
			comp.unloadContent();
			result = collections.remove(gc);
			collectionsToUpdate.remove(gc);
			if (Removed != null) {
				Removed.invoke(gc);
			}
		}
		return result;
	}

	public boolean removeAt(int idx) {
		if (isClear) {
			return false;
		}
		IGameComponent comp = collections.remove(idx);
		boolean result = (comp != null);
		if (result) {
			if (comp instanceof DrawableGameComponent) {
				((DrawableGameComponent) comp).unloadContent();
			}
			collectionsToUpdate.remove(comp);
			if (Removed != null) {
				Removed.invoke(comp);
			}
		}
		return result;
	}

	private ComponentEvent Added;

	private ComponentEvent Removed;

	public ComponentEvent getAdded() {
		return Added;
	}

	public void setAdded(ComponentEvent added) {
		Added = added;
	}

	public ComponentEvent getRemoved() {
		return Removed;
	}

	public void setRemoved(ComponentEvent removed) {
		Removed = removed;
	}
}
